include/ibamr/VCINSStaggeredHierarchyIntegrator.h

```diff
@@ -0,0 +1,655 @@
+                                               double new_time,
+                                               bool skip_synchronize_new_state_data,
+                                               int num_cycles = 1);
+
+    /*!
+     * Virtual method to regrid the patch hierarchy.
+     */
+    virtual void regridHierarchy();
+
+    /*!
+     * Explicitly remove nullspace components from a solution vector.
+     */
+    void removeNullSpace(const SAMRAI::tbox::Pointer<SAMRAI::solv::SAMRAIVectorReal<NDIM, double> >& sol_vec);
+
+    /*!
+     * Register a variable mass density variable with the hierarchy integrator.
+     */
+    void registerMassDensityVariable(SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > rho_var);
+
+    /*!
+     * Get the mass density variable registered with the hierarchy integrator.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > getMassDensityVariable() const;
+
+    /*!
+     * Register a variable viscosity variable with the hierarchy integrator.
+     */
+    void registerViscosityVariable(SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > mu_var);
+
+    /*!
+     * Get the viscosity variable registered with the hierarchy integrator.
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > getViscosityVariable() const;
+
+    /*!
+     * Set the interpolation type used for material properties rho
+     */
+    void setDensityVCInterpolationType(const IBTK::VCInterpType vc_interp_type);
+
+    /*!
+     * Set the interpolation type used for material properties mu
+     */
+    void setViscosityVCInterpolationType(const IBTK::VCInterpType vc_interp_type);
+
+    /*!
+     * \brief Function to reset fluid density or viscosity if they are
+     * maintained by this integrator.
+     */
+    typedef void (*ResetFluidPropertiesFcnPtr)(int property_idx,
+                                               SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > property_var,
+                                               SAMRAI::tbox::Pointer<IBTK::HierarchyMathOps> hier_math_ops,
+                                               int cycle_num,
+                                               double time,
+                                               double current_time,
+                                               double new_time,
+                                               void* ctx);
+
+    /*!
+     * \brief Register interface neighborhood locating functions.
+     */
+    void registerResetFluidDensityFcn(ResetFluidPropertiesFcnPtr callback, void* ctx);
+
+    /*!
+     * \brief Register interface neighborhood locating functions.
+     */
+    void registerResetFluidViscosityFcn(ResetFluidPropertiesFcnPtr callback, void* ctx);
+
+    /*!
+     * \brief Supply initial conditions for the density field, if maintained by the fluid integrator.
+     */
+    void registerMassDensityInitialConditions(SAMRAI::tbox::Pointer<IBTK::CartGridFunction> rho_init_fcn);
```

Comment:
I wonder if we should think about replacing a lot of uses of CartGridFunction with callbacks?